### PR TITLE
log operations

### PR DESCRIPTION
**Pull request by ayrat555**
yellow-paper

### CODE REVIEW

File: lib/evm.ex [hunk 1]
```diff
@@ -6,23 +6,28 @@ defmodule EVM do
...
+          address() => account()
         }
   @type trie_root :: MerklePatriciaTree.Trie.root_hash()
   @type val :: integer()
   @type address :: <<_::160>>
   @type hash :: <<_::256>>
   @type timestamp :: integer()
+  @type log_entry :: %{
+    address: address(),
+    topics: [],
```
Comment (hayesgm) [inline lib/evm.ex]: this is a list of what?

Comment (ayrat555) [reply lib/evm.ex]: @hayesgm it's list of integers. I'll add type, thanks. 

this pr is not finished yet. I encountered bug in `calldatacopy` operation - https://github.com/exthereum/evm/issues/46. That's why tests are failing.

Would you mind reviewing it one more time when it's finished? I hope I can finish it today or by monday in worst case scenario.

**Comment by ayrat555** [comment]
@hayesgm @masonforest Now it's finished! CR plz

File: lib/evm/gas.ex [hunk 2]
```diff
@@ -390,6 +390,26 @@ defmodule EVM.Gas do
...
+
+  def operation_cost(:log2, [_offset, size | _], _machine_state, _exec_end) do
+    @g_log + @g_logdata * size + @g_logtopic * 2
+  end
+
+  def operation_cost(:log3, [_offset, size | _], _machine_state, _exec_end) do
+    @g_log + @g_logdata * size + @g_logtopic * 3
+  end
+
+  def operation_cost(:log4, [_offset, size | _], _machine_state, _exec_end) do
```
Comment (masonforest) [inline lib/evm/gas.ex]: I feel like we could refactor these even more 🤔something like?:

    def operation_cost(:log4, [_offset, size | _], _machine_state, _exec_end) do
        log_cost(4)
    end

    defp log_cost(count)
        @g_log + @g_logdata * size + @g_logtopic * count
    end

Comment (ayrat555) [reply lib/evm/gas.ex]: I thought about it. But in Yellow Paper in cost function there are 4 clauses for each of log operations and I think it's more readable.

**Comment by masonforest** [comment]
👍

**Pull request merged**